tests/unit/test_service.py

```diff
@@ -38,3 +44,117 @@ def test_install_apt_package_gpg_key_error(
 
     with pytest.raises(RuntimeError, match="Error installing the agent service"):
         charm._on_install(MagicMock(spec=ops.InstallEvent))
+
+
+def test_on_install(harness: ops.testing.Harness, monkeypatch: pytest.MonkeyPatch):
+    """
+    arrange: Harness with mocked apt module.
+    act: run initial hook.
+    assert: The installation should pass without error and charm in blocked state.
+    """
+    apt_repository_mapping_mock = MagicMock()
+    apt_import_key_mock = MagicMock()
+    apt_update_mock = MagicMock()
+    apt_add_package_mock = MagicMock()
+
+    apt_repository_mapping_mock.__contains__.return_value = False
+    monkeypatch.setattr(apt, "RepositoryMapping", apt_repository_mapping_mock)
+    monkeypatch.setattr(apt, "import_key", apt_import_key_mock)
+    monkeypatch.setattr(apt, "update", apt_update_mock)
+    monkeypatch.setattr(apt, "add_package", apt_add_package_mock)
+
+    harness.begin_with_initial_hooks()
+    assert apt_add_package_mock.call_count == 2
+    assert harness.charm.unit.status.name == ops.BlockedStatus.name
+
+
+def test_restart_service(
+    harness: ops.testing.Harness,
+    monkeypatch: pytest.MonkeyPatch,
+    agent_relation_data: dict,
+    service_configuration_template: str,
+):
+    """
+    arrange: Harness with mocked systemd and fs-related methods.
+    act: add relation with jenkins-k8s with mock relation data and restart the agent service.
+    assert: Configuration file content should match the relation data and
+    restart should not raise any error and the charm should be in active state.
+    """
+    pathlib_write_text_mock = MagicMock()
+    monkeypatch.setattr(Path, "write_text", pathlib_write_text_mock)
+    monkeypatch.setattr(Path, "mkdir", MagicMock)
+    monkeypatch.setattr(os, "chmod", MagicMock)
+    monkeypatch.setattr(os, "chown", MagicMock)
+    monkeypatch.setattr(systemd, "daemon_reload", MagicMock)
+    monkeypatch.setattr(systemd, "service_restart", MagicMock)
+    monkeypatch.setattr(systemd, "service_running", MagicMock(return_value=True))
+    monkeypatch.setattr(os.path, "exists", MagicMock(return_value=True))
+
+    harness.add_relation(AGENT_RELATION, "jenkins-k8s", unit_data=agent_relation_data)
+    harness.begin()
+    charm: JenkinsAgentCharm = harness.charm
+    start_event_mock = MagicMock(spec=ops.StartEvent)
+    charm._on_start(start_event_mock)
```

Comment:
Can you test by triggering this events instead?

Reply:
yes, I updated the code to trigger the event